app/forms/date_form.py

```diff
@@ -84,3 +98,62 @@ class MonthYearDateForm(Form):
     MonthYearDateForm.month = SelectField(choices=month_choices, default='', validators=validate_with)
 
     return MonthYearDateForm
+
+
+def get_min_max_dates_for_single_date_period_validation(answer, answer_store, metadata):
+    """
+    Gets attributes within a minimum or maximum of a date field and validates that the entered date
+    is valid.
+
+    :param answer: The answer which contains the minimum or maximum
+    :param answer_store: The current answer store
+    :param metadata: metadata for reference meta dates
+    :return: attributes
+    """
+    minimum_referenced_date, maximum_referenced_date = None, None
+
+    if 'minimum' in answer:
+        minimum_referenced_date = get_referenced_offset_value(answer['minimum'], answer_store, metadata)
+    if 'maximum' in answer:
+        maximum_referenced_date = get_referenced_offset_value(answer['maximum'], answer_store, metadata)
+
+    if minimum_referenced_date and maximum_referenced_date:
+        if minimum_referenced_date > maximum_referenced_date:
+            raise Exception('The minimum offset date is greater than the maximum offset date for {}.'.format(answer['id']))
+
+    return minimum_referenced_date, maximum_referenced_date
+
+
+def get_referenced_offset_value(answer_min_or_max, answer_store, metadata):
+    """
+    Gets value of the referenced date type, whether it is a value,
+    id of an answer or a meta date. Then adds/subtracts offset from that value and returns
+    the new offset value
+
+    :param answer_min_or_max: The minimum or maximum object which contains
+    the referenced value.
+    :param answer_store: The current answer store
+    :param metadata: metadata for reference meta dates
+    :return: value
+    """
+    value = None
+
+    if 'value' in answer_min_or_max:
+        if answer_min_or_max['value'] == 'now':
+            value = datetime.utcnow().strftime('%Y-%m-%d')
+        else:
+            value = answer_min_or_max['value']
+    elif 'meta' in answer_min_or_max:
+        value = get_metadata_value(metadata, answer_min_or_max['meta'])
+    elif 'answer_id' in answer_min_or_max:
+        value = get_answer_store_value(answer_min_or_max['answer_id'], answer_store, group_instance=0)
+
+    if 'offset_by' in answer_min_or_max:
+        offset = answer_min_or_max['offset_by']
+        value = convert_to_datetime(value) + relativedelta(years=offset.get('years', 0),
+                                                           months=offset.get('months', 0),
+                                                           days=offset.get('days', 0))
+    else:
+        value = convert_to_datetime(value)
```

Comment:
`value = convert_to_datetime(value)` is repeated. This could be simplified to:

```
value = convert_to_datetime(value)

if 'offset_by' in answer_min_or_max:
    offset = answer_min_or_max['offset_by']
    value += relativedelta(years=offset.get('years', 0),
                           months=offset.get('months', 0),
                           days=offset.get('days', 0))
```